src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,151 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> tail;
+    private Node<T> head;
+    private int size;
+
+    public MyLinkedList() {
+        tail = null;
+        head = null;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        if (tail == null) {
+            tail = head = new Node<>(null, value, null);
+        } else if (index == 0) {
+            tail.prev = new Node<>(null, value, tail);
+            tail = tail.prev;
+        } else if (index == size) {
+            head.next = new Node<>(head, value, null);
+            head = head.next;
+        } else {
+            Node<T> current = tail;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            current.prev.next = new Node<>(current.prev, value, current);
+            current.prev = current.prev.next;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexOutOfBoundsException(index);
+        Node<T> current = tail;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexOutOfBoundsException(index);
+        Node<T> current = tail;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
```

Comment:
You should create the method `getNodeByIndex` or like this, and reuse it when you need to get node by index. [Don't forget about optimization](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa). Also, it's good idea to call method `isIndexOutOfBoundsException` inside it to get rid of line-by-line calling this 2 methods.